lib/ui/painting/codec.cc

```diff
@@ -79,10 +84,79 @@ static sk_sp<SkImage> DecodeImage(fml::WeakPtr<GrContext> context,
   }
 }
 
+static bool needsResize(const double currentWidth,
+                        const double currentHeight,
+                        const double targetWidth,
+                        const double targetHeight) {
+  if (fuzzyEquals(currentWidth, kDoNotResizeImage) ||
+      fuzzyEquals(currentHeight, kDoNotResizeImage)) {
+    return false;
+  }
+  if (fuzzyEquals(currentWidth, targetWidth) &&
+      fuzzyEquals(currentHeight, targetHeight)) {
+    return false;
+  }
+  return true;
+}
+
+static sk_sp<SkImage> DecodeAndResizeImageToExactSize(
+    fml::WeakPtr<GrContext> context,
+    SkImageInfo scaledImageInfo,
+    sk_sp<SkData> buffer,
+    size_t trace_id) {
+  TRACE_FLOW_STEP("flutter", kInitCodecTraceTag, trace_id);
+  TRACE_EVENT0("flutter", "DecodeAndResizeImageToExactSize");
+
+  SkBitmap bitmap = SkBitmap();
+  if (!bitmap.tryAllocPixels(scaledImageInfo)) {
+    FML_LOG(ERROR) << "Unable to allocate bitmap. Returning un-scaled image.";
+    return DecodeImage(context, buffer, trace_id);
+  }
+
+  // Do not create a cross context image here, since it can not be resized.
+  sk_sp<SkImage> image = SkImage::MakeFromEncoded(std::move(buffer));
+
+  if (image == nullptr || !image.get()) {
+    FML_LOG(ERROR) << "Failed to decode image.";
+    return nullptr;
+  }
+
+  if (!image->scalePixels(bitmap.pixmap(), kLow_SkFilterQuality)) {
+    FML_LOG(ERROR) << "Failed to scale pixels. Returning un-scaled image.";
+    return DecodeImage(context, buffer, trace_id);
+  }
+
+  // This indicates that we do not want a "linear blending" decode.
+  sk_sp<SkColorSpace> dstColorSpace = nullptr;
+  return SkImage::MakeCrossContextFromPixmap(context.get(), bitmap.pixmap(),
+                                             true, dstColorSpace.get(), true);
+}
+
+static sk_sp<SkImage> DecodeAndResizeImage(fml::WeakPtr<GrContext> context,
+                                           std::unique_ptr<SkCodec>& skCodec,
+                                           sk_sp<SkData> buffer,
+                                           const double targetWidth,
+                                           const double targetHeight,
+                                           size_t trace_id) {
+  const SkImageInfo imageInfo = skCodec->getInfo();
+
+  const int width = imageInfo.width();
+  const int height = imageInfo.height();
+
+  if (needsResize(width, height, targetWidth, targetHeight)) {
+    return DecodeAndResizeImageToExactSize(
+        context, imageInfo.makeWH(targetWidth, targetHeight), buffer, trace_id);
```

Comment:
Note that `makeWH` only takes integers. We probably should only allow integers to be sent to the API.